src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,165 @@
+    -- rightToLeft ('fromGraph' ('Algebra.Graph.vertex' (Left 1)))                              == Map.'Map.empty'
+    -- rightToLeft ('fromGraph' ('Algebra.Graph.vertex' (Right 1)))                             == Map.'Map.singleton' 2 Set.'Set.empty'
+    -- rightToLeft ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Right 2)))                      == Map.'Map.singleton' 2 (Set.'Set.singleton' 1)
+    -- rightToLeft ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Right 1)))                      == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- rightToLeft ('fromGraph' ('Algebra.Graph.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+    -- @
+    rightToLeft :: Map.Map b (Set.Set a)
+} deriving (Eq, Show)
+
+-- | Check that the internal graph representation is consistent, i.e. that all
+-- edges that are present in the 'leftToRight' map are present in the
+-- 'rightToLeft' map.
+-- Complexity: /O(m log m)/
+--
+-- @
+-- consistent ('fromGraph' 'Algebra.Graph.empty')      == True
+-- consistent ('fromGraph' ('Algebra.Graph.vertex' x)) == True
+-- consistent ('fromGraph' ('Algebra.Graph.edge' x y)) == True
+-- consistent ('fromGraph' ('Algebra.Graph.edges' xs)) == True
+-- consistent ('fromGraph' g)          == True
+-- consistent ('fromAdjacencyMap' am)  == True
+-- @
+consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
+consistent (BAM lr rl) = internalEdgeList lr == sort (map swap $ internalEdgeList rl)
+
+-- | The list of edges of a bipartite adjacency map
+--
+-- /Note: this function is for internal use only./
+internalEdgeList :: Map.Map a (Set.Set b) -> [(a, b)]
+internalEdgeList lr = [ (u, v) | (u, vs) <- Map.toAscList lr, v <- Set.toAscList vs ]
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y ('fromGraph' 'Algebra.Graph.empty')                                 == False
+-- hasEdge x y ('fromGraph' ('Algebra.Graph.edge' (Left x) (Right y)))             == True
+-- hasEdge 1 2 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Left 2)))              == False
+-- hasEdge 2 3 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Right 2)))             == False
+-- hasEdge x y ('fromGraph' ('Algebra.Graph.Overlay' g ('Algebra.Graph.edge' (Left x) (Right y)))) == True
+-- @
+hasEdge :: (Ord a, Ord b) => a -> b -> AdjacencyMap a b -> Bool
+hasEdge u v (BAM m _) = ((Set.member v) <$> (u `Map.lookup` m)) == Just True
+
+-- | Adds all edges needed to make the graph undirected.
+-- Complexity: /O(m log n)/
+--
+-- /Note: this function is for internal use only./
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
+addReverseEdges m = AM.overlay m $ AM.edges [ (v, u) | (u, v) <- AM.edgeList m ]
+
+-- | Constructs a bipartite 'Bipartite.AdjacencyMap' from
+-- 'Algebra.Graph.AdjacencyMap' with given part identifiers, adding all needed
```

Comment:
Here please use double quotes to link to the module instead (otherwise it may be confusing to see two equally named links).